DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/task_sequences/README.md

```diff
@@ -0,0 +1,302 @@
+
+## Key Implementation Details
+
+This sample demonstrates some key concepts:
+- How to execute multiple dependent loops in parallel using the `task_sequence` class, `pipe`s and the `async()` function call
+- How the compiler automatically adds depth to pipes that bypass one or more loops
+
+The demonstration system in this tutorial is structured like this:
+
+![](assets/LoopStructure.svg)
+
+When you write code containing a sequence of loops, each loop in the sequence must wait for the previous loop in the sequence to fully complete. This results in long execution time and poor occupancy of the FPGA datapath. Observe in the following graphic that it takes many hundreds of clock cycles before the output pipe produces any output:
+
+![](assets/simulation_naive.png)
+
+Task sequences allow you to schedule loops to run in parallel, so data quickly flows through sequences of loops. A consumer loop can process an output as soon as a producer loop produces it, without waiting for the whole producer loop to finish executing. Observe in the following graphic that the output pipe produces results almost immediately:
+
+![](assets/simulation_task_sequence.png)
+
+>**Note**: This is only possible when the loops process data in the same order. That is, the first piece of data the second loop consumes should be the first piece of data produced by the first loop. Converting loops to task sequences and executing them in parallel will not help performance if the first piece of data consumed by the second loop is the last piece of data produced by the first loop, because then the second loop requires the first loop to completely finish.
+
+### Code Example 
+
+To use a task sequence in your design, include the `<sycl/ext/intel/experimental/task_sequence.hpp>` header file in your source code. The `task_sequence` class is a templated class with 3 parameters:
+
+| Template Parameter | Type     | Default Value | Description
+|--------------------|----------|---------------|---
+| Task function      | callable | N/A           | A callable object `f` that defines the asynchronous task to be associated with the `task_sequence`. The callable object `f` must meet the following requirements: <br> • The object `f` must be statically resolvable at compile time, which means it is not a function pointer. <br> • The object `f` must not be an overloaded function. <br> • The return type (`ReturnT`) and argument types (`ArgsT…`) of object f must be resolvable and fixed.
+| Invocation Capacity* | `uint32_t` | 1 | The size of the hardware queue instantiated for `async()` function calls. This parameter value corresponds to the minimum number of outstanding `async()` function calls to be supported. When the outstanding number of `async()` function calls reaches this value, further calls may block until the number of outstanding calls is reduced to the `invocation_capacity`. The default value of this parameter is 1.
```

Comment:
After reading the description of these two parameters (invocation capacity and response capacity), users will naturally wonder how to select appropriate values for them. We can introduce some basic principles for choosing these parameters.

Also, I would like to add some diagram illustrating the effect of using these two parameters. 

For example, for the invocation capacity, user might want to specify a larger value if they try to issue `async` calls very frequently, and then before they `get` all the results from the task function, they want the caller to move onto some other tasks. If we do not specify enough invocation capacity, the task function may stall the upstream caller so that the caller will block and stops making progress.
![invoc_capacity](https://github.com/chanjunc-intel/oneAPI-samples/assets/144172047/86ab857e-8264-48ca-8966-3e57d92570e9)

Similarly, specifying the correct response capacity creates the buffer at the `get` end so that the `get` calls will not block. A case to illustrate this might be when a user more frequently issues `async` and occasionally issues `get`. We don't want the task function to block because there is no consumer consuming its output (since `get` only happens after multiple `async`) - instead, we just want the results to be buffered and awaits for `get` to consume so that the task function makes progress. For example, if we don't have enough response capacity, then before the result being consumed, the task will backpressure upstream.
![response_capacity](https://github.com/chanjunc-intel/oneAPI-samples/assets/144172047/91c205ad-6a59-4283-8950-d2746a03d49f)

Reply:
I think this is important material, but it belongs in the reference manual and the 'resource sharing' code sample. This issue will only turn up if you have a re-entrant kernel (that is, a kernel that is invoked multiple times externally) that `async()`s and `get()`s a task sequence once each invocation.

Reply:
Understood :)